File: app/api/v1/endpoints/shareplace.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, desc, func
from typing import Optional
from app.core.database import get_db
from app.models.tl_file import TLFile, AuthStatus

router = APIRouter()


@router.get("")
async def get_shareplace(
    genre: Optional[str] = Query(None),
    country: Optional[str] = Query(None),
    file_type: Optional[str] = Query(None),
    sort: str = Query("pulse"),
    limit: int = Query(20, le=50),
    offset: int = Query(0),
    db: AsyncSession = Depends(get_db),
):
    q = select(TLFile).where(
        TLFile.auth_status == AuthStatus.verified,
        TLFile.shared == True,
        TLFile.revenue_held == False,
    )
    if genre:
        q = q.where(TLFile.genre == genre)
    if country:
        q = q.where(TLFile.country == country)
    if file_type:
        q = q.where(TLFile.file_type == file_type)
    if sort == "pulse":
        q = q.order_by(desc(TLFile.pulse))
    elif sort == "new":
        q = q.order_by(desc(TLFile.created_at))
    elif sort == "tl":
        q = q.order_by(desc(TLFile.file_tl))
    q = q.offset(offset).limit(limit)
    result = await db.execute(q)
    files = result.scalars().all()
    return [
        {
            "id": f.id,
            "title": f.title,
            "artist": f.artist,
            "genre": f.genre,
            "country": f.country,
            "file_type": f.file_type,
            "file_tl": f.file_tl,
            "pulse": f.pulse,
            "play_count": f.play_count,
            "revenue": f.revenue,
            "auth_status": f.auth_status,
            "created_at": f.created_at,
        }
        for f in files
    ]


@router.get("/contributor-ranking")
async def get_contributor_ranking(
    period: str = Query("weekly"),
    content_type: Optional[str] = Query(None),
    limit: int = Query(20, le=50),
    db: AsyncSession = Depends(get_db),
):
    from datetime import datetime, timedelta
    from sqlalchemy import Float as SAFloat
    from app.models.user import User

    q = (
        select(
            TLFile.user_id,
            func.sum(TLFile.revenue).label("total_revenue"),
            func.sum(TLFile.pulse).label("total_pulse"),
            func.sum(TLFile.play_count).label("total_plays"),
            func.count(TLFile.id).label("verified_tracks"),
        )
        .where(
            TLFile.auth_status == AuthStatus.verified,
            TLFile.shared == True,
        )
        .group_by(TLFile.user_id)
    )

    if period == "weekly":
        cutoff = datetime.utcnow() - timedelta(days=7)
        q = q.where(TLFile.updated_at >= cutoff)
    elif period == "monthly":
        cutoff = datetime.utcnow() - timedelta(days=30)
        q = q.where(TLFile.updated_at >= cutoff)

    q = q.order_by(desc("total_pulse") if period != "alltime" else desc("total_revenue"))
    q = q.limit(limit)

    result = await db.execute(q)
    rows = result.all()

    ranking = []
    for i, row in enumerate(rows):
        user_result = await db.execute(select(User).where(User.id == row.user_id))
        user = user_result.scalar_one_or_none()
        if not user:
            continue
        ranking.append({
            "rank": i + 1,
            "user_id": row.user_id,
            "username": user.username,
            "poc_index": user.poc_index,
            "total_revenue": row.total_revenue or 0,
            "total_pulse": row.total_pulse or 0,
            "total_plays": row.total_plays or 0,
            "verified_tracks": row.verified_tracks or 0,
            "false_dispute_strikes": user.false_dispute_strikes,
            "account_forfeited": user.account_forfeited,
        })
    return ranking

File: src/routes/shareplace.ts
import { Hono } from 'hono';
import type { Env } from '../types';

const router = new Hono<{ Bindings: Env }>();

// GET / - SharePlace 목록
router.get('/', async (c) => {
  const genre = c.req.query('genre');
  const fileType = c.req.query('file_type');
  const sort = c.req.query('sort') || 'pulse';
  const limit = Number(c.req.query('limit') || 20);
  const offset = Number(c.req.query('offset') || 0);

  let query = `SELECT f.*, u.username FROM tl_files f
    JOIN users u ON f.user_id = u.id
    WHERE f.auth_status = 'verified' AND f.shared = 1 AND f.revenue_held = 0`;
  const params: (string | number)[] = [];

  if (genre) { query += ' AND f.genre = ?'; params.push(genre); }
  if (fileType) { query += ' AND f.file_type = ?'; params.push(fileType); }

  const orderCol = sort === 'new' ? 'f.created_at' : sort === 'tl' ? 'f.file_tl' : 'f.pulse';
  query += ` ORDER BY ${orderCol} DESC LIMIT ? OFFSET ?`;
  params.push(limit, offset);

  const rows = await c.env.DB.prepare(query).bind(...params).all();
  return c.json(rows.results);
});

// GET /contributor-ranking
router.get('/contributor-ranking', async (c) => {
  const period = c.req.query('period') || 'weekly';
  const limit = Number(c.req.query('limit') || 20);

  let dateFilter = '';
  if (period === 'weekly') dateFilter = `AND f.updated_at >= datetime('now', '-7 days')`;
  else if (period === 'monthly') dateFilter = `AND f.updated_at >= datetime('now', '-30 days')`;

  const query = `
    SELECT
      f.user_id,
      u.username,
      u.poc_index,
      u.false_dispute_strikes,
      u.account_forfeited,
      SUM(f.revenue) as total_revenue,
      SUM(f.pulse) as total_pulse,
      SUM(f.play_count) as total_plays,
      COUNT(f.id) as verified_tracks
    FROM tl_files f
    JOIN users u ON f.user_id = u.id
    WHERE f.auth_status = 'verified' AND f.shared = 1 ${dateFilter}
    GROUP BY f.user_id
    ORDER BY total_pulse DESC
    LIMIT ?
  `;

  const rows = await c.env.DB.prepare(query).bind(limit).all();
  const results = (rows.results as any[]).map((r, i) => ({ ...r, rank: i + 1 }));
  return c.json(results);
});

export default router;
